### ISSUE TITLE
Debugging/Investigating `hsp_hydr_conversion.R` for use in hspf meta_model

### ISSUE THREAD

**Issue by glenncampagna** [HARPgroup/HARParchive#778]
I started by running a test with Rob's most recent meta_model updates on the hspf branch with the run command:
`~/meta_model/run_model hspf_cbp6 vadeq_2021 JL1_6770_6850 auto river`
And I _did_ get a daily hydr csv created:
```
-rw-rw-r-- 1 gcampagna11 allmodelers  48M Mar 24 17:50 JL1_6770_6850_hydr.csv
-rw-rw-r-- 1 gcampagna11 allmodelers 1.4M Mar 24 17:50 JL1_6770_6850_hydrd_wy.csv
```
but it looked like this:
```
                index    OVOL3     Qout       date hour day month year
1: 1984-10-01 11:30:00 15.59937 188.7524 1984-10-01 11.5   1    10 1984
2: 1984-10-02 11:30:00 11.56653 139.9550 1984-10-02 11.5   2    10 1984
3: 1984-10-03 11:30:00 10.05994 121.7252 1984-10-03 11.5   3    10 1984
   divr_cfs diva_cfs ps_afd wd_mgd ps_mgd demand_mgd Qbaseline wd_imp_child_mgd
1:       NA       NA     NA     NA     NA         NA        NA                0
2:       NA       NA     NA     NA     NA         NA        NA                0
3:       NA       NA     NA     NA     NA         NA        NA                0
   wd_cumulative_mgd ps_cumulative_mgd ps_nextdown_mgd
1:                NA                NA               0
2:                NA                NA               0
3:                NA                NA               0
```
For reference, the hourly hydr csv for the same run looks like this: 
   ```
              index    OVOL3     Qout       date hour day month year
1: 1984-01-01 00:00:00 26.72870 323.4173 1984-01-01    0   1     1 1984
2: 1984-01-01 01:00:00 23.03203 278.6876 1984-01-01    1   1     1 1984
3: 1984-01-01 02:00:00 20.02072 242.2507 1984-01-01    2   1     1 1984
   divr_cfs diva_cfs    ps_afd wd_mgd    ps_mgd demand_mgd Qbaseline
1: 3.122533        0 0.3715839     NA 0.1210806   2.018183        NA
2: 3.122533        0 0.3715839     NA 0.1210806   2.018183        NA
3: 3.122533        0 0.3715839     NA 0.1210806   2.018183        NA
   wd_imp_child_mgd wd_cumulative_mgd ps_cumulative_mgd ps_nextdown_mgd
1:                0                NA         0.1210806               0
2:                0                NA         0.1210806               0
3:                0                NA         0.1210806               0

```
However, this was with the old-version script in the old location indicated by 
`Rscript $MODEL_ROOT/run/export/hsp_hydr_conversion.R $csvfile`

- [x] Test with new-location Rscript `meta_model/scripts`
- [x] Hourly data is aggregated twice in `hsp_hydr_conversion.R` which I think is a mistake ... fix this & test or figure out why it's needed

**Comment by glenncampagna** [comment]
Another test with the line `Rscript $META_MODEL_ROOT/scripts/river/hsp_hydr_conversion.R $csvfile` to indicate the correct R script to be run
The hourly hydr csv:
```
                 index    OVOL3     Qout       date hour day month year
1: 1984-01-01 00:00:00 26.72870 323.4173 1984-01-01    0   1     1 1984
2: 1984-01-01 01:00:00 23.03203 278.6876 1984-01-01    1   1     1 1984
3: 1984-01-01 02:00:00 20.02072 242.2507 1984-01-01    2   1     1 1984
   divr_cfs diva_cfs    ps_afd wd_mgd    ps_mgd demand_mgd Qbaseline
1: 3.122533        0 0.3715839     NA 0.1210806   2.018183        NA
2: 3.122533        0 0.3715839     NA 0.1210806   2.018183        NA
3: 3.122533        0 0.3715839     NA 0.1210806   2.018183        NA
   wd_imp_child_mgd wd_cumulative_mgd ps_cumulative_mgd ps_nextdown_mgd
1:                0                NA         0.1210806               0
2:                0                NA         0.1210806               0
3:                0                NA         0.1210806               0
```
The daily hydr csv:
  ```
              index    OVOL3     Qout       date hour day month year
1: 1984-10-01 11:30:00 15.59937 188.7524 1984-10-01 11.5   1    10 1984
2: 1984-10-02 11:30:00 11.56653 139.9550 1984-10-02 11.5   2    10 1984
3: 1984-10-03 11:30:00 10.05994 121.7252 1984-10-03 11.5   3    10 1984
   divr_cfs diva_cfs ps_afd wd_mgd ps_mgd demand_mgd Qbaseline wd_imp_child_mgd
1:       NA       NA     NA     NA     NA         NA        NA                0
2:       NA       NA     NA     NA     NA         NA        NA                0
3:       NA       NA     NA     NA     NA         NA        NA                0
   wd_cumulative_mgd ps_cumulative_mgd ps_nextdown_mgd
1:                NA                NA               0
2:                NA                NA               0
3:                NA                NA               0

```
- There is a loss of data between the hourly and daily hydr files in columns `divr`, `ps`, `demand`, and more
- The only columns that I think are okay with no data currently are `wd` columns and `Qbaseline`, since the columns to find withdrawal aren't present in the hspf hydr csv and Qbaseline depends on wd

**Comment by glenncampagna** [comment]
### Source of data loss: aggregation from hourly to daily 
Manual testing of `hsp_hydr_conversion.R` in terminal R:
```
head(hydr,3)
                 index    OVOL3     Qout       date hour day month year
1: 1984-01-01 00:00:00 26.72870 323.4173 1984-01-01    0   1     1 1984
2: 1984-01-01 01:00:00 23.03203 278.6876 1984-01-01    1   1     1 1984
3: 1984-01-01 02:00:00 20.02072 242.2507 1984-01-01    2   1     1 1984
   divr_cfs diva_cfs    ps_afd wd_mgd    ps_mgd demand_mgd Qbaseline
1: 3.122533        0 0.3715839     NA 0.1210806   2.018183        NA
2: 3.122533        0 0.3715839     NA 0.1210806   2.018183        NA
3: 3.122533        0 0.3715839     NA 0.1210806   2.018183        NA
> hydrd = aggregate(hydr, list(hydr$date), 'mean')
> head(hydrd,3)
     Group.1               index     OVOL3      Qout       date hour day month
1 1984-01-01 1984-01-01 11:30:00 11.913753 144.15641 1984-01-01 11.5   1     1
2 1984-01-02 1984-01-02 11:30:00  7.119201  86.14233 1984-01-02 11.5   2     1
3 1984-01-03 1984-01-03 11:30:00  6.772316  81.94503 1984-01-03 11.5   3     1
  year divr_cfs diva_cfs    ps_afd wd_mgd    ps_mgd demand_mgd Qbaseline
1 1984 3.122533        0 0.3715839     NA 0.1210806   2.018183        NA
2 1984       NA       NA        NA     NA        NA         NA        NA
3 1984       NA       NA        NA     NA        NA         NA        NA
  wd_imp_child_mgd wd_cumulative_mgd ps_cumulative_mgd ps_nextdown_mgd
1                0                NA         0.1210806               0
2                0                NA                NA               0
3                0                NA                NA               0
```

**Comment by glenncampagna** [comment]
### More findings - 24th/0th hour difference is cause of data loss in hspf daily hydr csv 
Here's data from an hspf point source csv `ps_sep_div_ams_vadeq_2021_JL1_6770_6850_3000.csv`
```
      V1 V2 V3 V4        V5
 1: 1984  1  1  0 0.3715839
 2: 1984  1  1  1 0.3715839
 3: 1984  1  1  2 0.3715839
 4: 1984  1  1  3 0.3715839
 5: 1984  1  1  4 0.3715839
 6: 1984  1  1  5 0.3715839
 7: 1984  1  1  6 0.3715839
 8: 1984  1  1  7 0.3715839
 9: 1984  1  1  8 0.3715839
10: 1984  1  1  9 0.3715839
11: 1984  1  1 10 0.3715839
12: 1984  1  1 11 0.3715839
13: 1984  1  1 12 0.3715839
14: 1984  1  1 13 0.3715839
15: 1984  1  1 14 0.3715839
16: 1984  1  1 15 0.3715839
17: 1984  1  1 16 0.3715839
18: 1984  1  1 17 0.3715839
19: 1984  1  1 18 0.3715839
20: 1984  1  1 19 0.3715839
21: 1984  1  1 20 0.3715839
22: 1984  1  1 21 0.3715839
23: 1984  1  1 22 0.3715839
24: 1984  1  1 23 0.3715839
25: 1984  1  1 24 0.3715839
26: 1984  1  2  1 0.3715839
27: 1984  1  2  2 0.3715839
```
Note the use of a _24th_ hour and no _0th_ hour, however hsp2 files have hours 0 and 23 but _not_ 24
These lines in join_col where hours of the hydr csv are matched/joined to hours of the hspf hourly wdm-format csv are what I believe is causing the data loss:
```
if (timestep == 'hour') {
  df1_joined <- sqldf(
    paste0("SELECT a.*, b.'", old_col,"' AS '", new_col, "'
        FROM df1 AS a 
         LEFT OUTER JOIN df2 AS b ON (a.year = b.year AND a.month = b.month AND a.day = b.day AND a.hour = b.hour)
         ORDER BY a.year,a.month,a.day,a.hour"
    )
  ) 
}
```
specifically, where `a.hour = b.hour`

The hspf hydr table is created in `create_hspf_hydr.R` with these lines specifying the time/hour range:
`hspf$index <- seq(mdy_hm("1/1/1984 0:00"), mdy_hm("12/31/2020 23:00"), by = "hour")`
where it seems like the sequence function automatically uses a 0th hour instead of 24th

The only solution to this I can think of at the moment is to convert all `24` values to `0` in the hour columns of the hspf csvs that are joined to the hydr table... an additional argument could even be added to `join_col.R`, maybe `hsp2/hspf`, to indicate if this hour conversion should happen

**Comment by glenncampagna** [comment]
### Side-by-Side Visual of what's causing the problem:
![hspfVsHsp2_hour:dayScreenshot](https://user-images.githubusercontent.com/104520197/228653532-a28b177b-0798-4293-b50d-1f9e007500ef.png)

**Comment by glenncampagna** [comment]
Thanks @rburghol, makes sense.
I did confirm that the `_0111.csv` which the hspf hydr table is created from does have the same timestep format as the other hspf csvs, with the 24th hours.
However I'm having trouble with the 24th hours always being converted to 0, even with that new line of code:
```
> index <- as.POSIXct(make_datetime(hspf$year,hspf$month,hspf$day,hspf$hour))
> head(index,50)
 [1] "1984-01-01 00:00:00 UTC" "1984-01-01 01:00:00 UTC"
 [3] "1984-01-01 02:00:00 UTC" "1984-01-01 03:00:00 UTC"
 [5] "1984-01-01 04:00:00 UTC" "1984-01-01 05:00:00 UTC"
 [7] "1984-01-01 06:00:00 UTC" "1984-01-01 07:00:00 UTC"
 [9] "1984-01-01 08:00:00 UTC" "1984-01-01 09:00:00 UTC"
[11] "1984-01-01 10:00:00 UTC" "1984-01-01 11:00:00 UTC"
[13] "1984-01-01 12:00:00 UTC" "1984-01-01 13:00:00 UTC"
[15] "1984-01-01 14:00:00 UTC" "1984-01-01 15:00:00 UTC"
[17] "1984-01-01 16:00:00 UTC" "1984-01-01 17:00:00 UTC"
[19] "1984-01-01 18:00:00 UTC" "1984-01-01 19:00:00 UTC"
[21] "1984-01-01 20:00:00 UTC" "1984-01-01 21:00:00 UTC"
[23] "1984-01-01 22:00:00 UTC" "1984-01-01 23:00:00 UTC"
[25] "1984-01-02 00:00:00 UTC" "1984-01-02 01:00:00 UTC"
[27] "1984-01-02 02:00:00 UTC" "1984-01-02 03:00:00 UTC"

```
even though the original data looks like this:
```
> head(hspf,50)
    year month day hour     ROVOL      Qout
 1: 1984     1   1    0 26.728704 323.41732
 2: 1984     1   1    1 23.032030 278.68756
 3: 1984     1   1    2 20.020719 242.25071
 4: 1984     1   1    3 17.585709 212.78707
 5: 1984     1   1    4 15.777524 190.90804
 6: 1984     1   1    5 14.469640 175.08264
 7: 1984     1   1    6 13.381302 161.91375
 8: 1984     1   1    7 12.456387 150.72229
 9: 1984     1   1    8 11.669604 141.20221
10: 1984     1   1    9 10.999783 133.09737
11: 1984     1   1   10 10.429261 126.19406
12: 1984     1   1   11  9.943154 120.31217
13: 1984     1   1   12  9.528883 115.29948
14: 1984     1   1   13  9.175900 111.02840
15: 1984     1   1   14  8.875283 107.39093
16: 1984     1   1   15  8.619449 104.29533
17: 1984     1   1   16  8.401700 101.66057
18: 1984     1   1   17  8.215671  99.40961
19: 1984     1   1   18  8.056107  97.47889
20: 1984     1   1   19  7.918974  95.81959
21: 1984     1   1   20  7.800865  94.39047
22: 1984     1   1   21  7.698892  93.15659
23: 1984     1   1   22  7.610602  92.08829
24: 1984     1   1   23  7.533919  91.16042
25: 1984     1   1   24  7.467080  90.35167
26: 1984     1   2    1  7.408592  89.64396
27: 1984     1   2    2  7.357188  89.02197

```

**Comment by glenncampagna** [comment]
### Possible work-around for timestamp issue in hspf, modification to `join_col.R` with arg for `hspf`
- Since we've found that the `lubridate` functions tried above convert the 24th hours of hspf csvs to 0th hours of the next day, I thought that we could do this same conversion to the day & hour columns of the hspf csvs in the same script they are joined with
- Creating an index of class date with lubridate, which will have the same hour & day sequencing as the hydr csv, and then re-generating the hour & day columns from that index would then have all timestamps and column values matching 

Here's what I used:
```
# index/timestamp adjustment if data to be joined is hspf
if (mod == 'hspf') {
  df2$index <- as.POSIXct(make_datetime(df2$year,df2$month,df2$day,df2$hour))
  df2$date <- as.Date(df2$index, origin = origin, tz = "UTC") 
  df2$hour <- hour(df2$index)
  df2$day <- day(df2$index)
  df2$month <- month(df2$index)
  df2$year <- year(df2$index)
}
```
Then, the join works as designed because everything lines up
I have applied these changes to the HARParchive version of this script for easy reversal

**Comment by glenncampagna** [comment]
### Test with the changes from above comment successful for meta-model
I ran a meta-model test using the version of the joins script with changes mentioned above and produced this daily hydr csv:

```
-rw-rw-r-- 1 gcampagna11 allmodelers  49M Mar 31 22:32 JL1_6770_6850_hydr.csv
-rw-rw-r-- 1 gcampagna11 allmodelers 2.1M Mar 31 22:33 JL1_6770_6850_hydrd_wy.csv

```
```
> head(hydrd,5)
                 index     OVOL3     Qout       date hour day month year
1: 1984-10-01 11:30:00 15.599370 188.7524 1984-10-01 11.5   1    10 1984
2: 1984-10-02 11:30:00 11.566530 139.9550 1984-10-02 11.5   2    10 1984
3: 1984-10-03 11:30:00 10.059940 121.7252 1984-10-03 11.5   3    10 1984
4: 1984-10-04 11:30:00  9.277578 112.2587 1984-10-04 11.5   4    10 1984
5: 1984-10-05 11:30:00  8.702687 105.3025 1984-10-05 11.5   5    10 1984
   divr_cfs diva_cfs    ps_afd wd_mgd    ps_mgd demand_mgd Qbaseline
1: 0.220788        0 0.3715839     NA 0.1210806  0.1427017        NA
2: 0.220788        0 0.3715839     NA 0.1210806  0.1427017        NA
3: 0.220788        0 0.3715839     NA 0.1210806  0.1427017        NA
4: 0.220788        0 0.3715839     NA 0.1210806  0.1427017        NA
5: 0.220788        0 0.3715839     NA 0.1210806  0.1427017        NA
   wd_imp_child_mgd wd_cumulative_mgd ps_cumulative_mgd ps_nextdown_mgd
1:                0                NA         0.1210806               0
2:                0                NA         0.1210806               0
3:                0                NA         0.1210806               0
4:                0                NA         0.1210806               0
5:                0                NA         0.1210806               0

```
so I guess this method works for what I tried to overcome but I'm not sure about the robustness of this fix..

**Comment by rburghol** [comment]
Great stuff @glenncampagna -- this is a bit tricky!  I had assumed that he zero hour format would result in just a guess of yesterday or today, not dropping of an hour entirely. 

Can you map out what *is* happening, and what we *think* should happen, so that we can be very deliberate about whether the proposed fix does what we want?  In other words, since "hspf" wdm export shows 1-24, 
- am I correct that the simple lubridate fix I proposed effectively drops an hour, so, not an accurate fix (by itself)?
- how do we be certain that the *correct* fix is converting 24 to 0, as opposed to subtracting 1 from each hour?

in other words, I think the *right* fix IS to do as you suggest, since hour 24 really is the zero hour, like when we stay up past midnight, we say "it's 12:15", we don't say "it's 00:15 in the morning", but that's kinda crazy since we are saying the 24th hour of today is actually tomorrow... but I want to just spell it out a touch to avoid misunderstanding and introducing a 1 hour time shift into the data.

**Comment by glenncampagna** [comment]
It seems like we are on the same page @rburghol. To clarify:
- The hspf csvs have a 24th hour of every day, while the hsp2 index has a 0 hour instead which starts the next day
- The lubridate fix you proposed simply changes the 24th hours to 0s of the next day (also adding 1 to the day of that row). I think this is the right think to do: a) It makes sense, like you explained. The next day starts at midnight (0:00). b) It's how the hsp2 index is formatted, so it makes things match up and makes things lubridate-friendly
- Since both hspf and hsp2 csvs start at 0:00 and the corresponding hours all match up except for the 24/0 hour lines, of course, I don't think subtraction would be necessary
- If we decide to do this conversion to make the timestamps match like I tested, I think it might be best as a new script, maybe `time_convert_hspf.R` ? `join_col.R` already has 6 arguments and works as designed, so a simple script that does the conversion and requires a couple of arguments could be easiest for future use

**Comment by rburghol** [comment]
@glenncampagna - thanks for all the clarifications.  I think a separate script is in order, let's called `time_fix.R` since this really is a bug in the wdm export routine IMO.  However, I have a second idea to explore.  The program `wdmtoolbox`, which we use in `/opt/model/meta_model/models/hsp2_cbp6/river/run/10_hydr_to_wdm` has a function called `extract`, which exports data from a WDM in a 2-column format, `ISO time,value`.

Example:
- `wdmtoolbox extract PM7_4581_4580.wdm,11`
- outputs:
```
2020-12-30 00:00:00,6.333595275878906
2020-12-30 01:00:00,6.307774543762207
2020-12-30 02:00:00,6.282342910766602
2020-12-30 03:00:00,6.257291316986084
2020-12-30 04:00:00,6.232605934143066
2020-12-30 05:00:00,6.208277225494385
2020-12-30 06:00:00,6.184295177459717
2020-12-30 07:00:00,6.160645484924316
2020-12-30 08:00:00,6.137323379516602
2020-12-30 09:00:00,6.114315509796143
2020-12-30 10:00:00,6.091614246368408
2020-12-30 11:00:00,6.069211959838867
2020-12-30 12:00:00,6.04709529876709
2020-12-30 13:00:00,6.0252604484558105
2020-12-30 14:00:00,6.003695964813232
2020-12-30 15:00:00,5.982397079467773
2020-12-30 16:00:00,5.961352825164795
2020-12-30 17:00:00,5.94055700302124
2020-12-30 18:00:00,5.920004844665527
2020-12-30 19:00:00,5.899688243865967
2020-12-30 20:00:00,5.879596710205078
2020-12-30 21:00:00,5.8597283363342285
2020-12-30 22:00:00,5.840074062347412
2020-12-30 23:00:00,5.8206305503845215
2020-12-31 00:00:00,5.801388263702393
2020-12-31 01:00:00,5.782345294952393
2020-12-31 02:00:00,5.763494968414307

```
